### ISSUE TITLE
Removing communications from ParMesh constructor from serial mesh

### ISSUE THREAD

**Issue by sebastiangrimberg** [mfem/mfem#2178]
I'm running into a similar issue as #934 and #1549 where loading the serial mesh on all MPI processors at the start of a simulation leads to an unfortunately large spike in memory usage before dropping after the serial mesh is deleted on all ranks after returning from the ParMesh constructor.

I know that some of the proposed solutions in the aforementioned issues have been taken as an internal project since it involves a good bit of work, but one perhaps quick solution I thought of would be to just stagger the serial mesh read and ParMesh constructors on all MPI ranks sharing the same node. This would clearly reduce scalability since processes would wait to load the mesh but could get around the memory limitations imposed by all processors reading the serial mesh simultaneously.

This currently seems possible for linear meshes since the ParMesh constructor is completely asynchronous in this case. However, for curved meshes, the construction of the ParGridFunction for the nodes from the global serial GridFunction has a couple of blocking global MPI calls which prevents this idea from being feasible.

I haven't thought this all the way through and maybe this is a shot in the dark, but has anyone thought about this idea and perhaps any solution for it? It seems like it would be possible to construct the ParGridFunction in a way that doesn't require any global communication so that the construction of the ParMesh from the serial mesh on each process could be done independently.

**Comment by sebastiangrimberg** [comment]
Adding to this based on some progress in case it is helpful to anyone else or anyone has comments/feedback:

I've implemented this approach and it seems to work (for conforming, non-NURBS meshes) with the following sequence: read the serial mesh and if it is curved, use something like `Mesh::SwapNodes` to delete the higher-order info from the mesh, saving the `GridFunction` only on the root process. Then create the ParMesh from the sequential mesh using the usual constructor and delete the serial mesh. This process can happen totally asynchronously so all ranks sharing a node can proceed sequentially or in batches so that the node memory is never overrun by too many simultaneous reads of the serial mesh.

Once all processes finish this, a simple `ParMesh::SetCurvature` on all processes to initialize the `ParGridFunction` for the nodes using the same parameters as the original serial mesh, followed by an`MPI_Scatterv` from the root to distribute the `GridFunction` vdofs from the serial mesh sets up the curvature information.

In my testing, this seems to be correct and allows me to get around the memory limits. It is obviously slower since some processes are waiting around while others read in the serial mesh from disk but when this is not a bottleneck on the total simulation time it makes little impact. As a final note, the changes in the pmesh-constr branch seem like they could be combined with this to fully unify the process of reading the serial mesh on on process and distributing all the information, but the code there is specialized for quad/hex meshes right now so needs a bit of work to be fully general.

**Comment by v-dobrev** [comment]
Hi @sebastiangrimberg,

This is interesting and may be useful for others. If I understand you correctly, your approach does not require changes in the MFEM classes -- is that correct? If that is the case, would you like to contribute your solution to the library? For example, we can add it as a "name constructor" in class ParMesh, similar to `ParMesh::MakeRefined`.

Thanks!

**Comment by sebastiangrimberg** [comment]
You're right @v-dobrev, this operates outside of the library and is just a change in my application code using the existing `Mesh` and `ParMesh` constructors. I'd be happy to contribute it, let me take some time to work out any issues/improvements before getting to that stage. Just wanted to follow up here in case anyone had any input or it could be useful in the meantime.

Thank you!

**Comment by sebastiangrimberg** [comment]
An update on this:

Even for non-curved meshes, there are cases where the ParMesh constructor still is blocking because of the GroupTopology creation [here](https://github.com/mfem/mfem/blob/8c27e5fb36d2f9f300c6e230d87e92001525d7c2/mesh/pmesh.cpp#L218), so my previous suggestion of a working approach doesn't work in general. Open to suggestions on possible remedies or improvements.

**Comment by v-dobrev** [comment]
Unfortunately, I have not had time to look at this for possible solutions for the GroupTopology creation issue.

**Comment by sebastiangrimberg** [comment]
I have a temporary solution which I'm not sure is worth making a PR for, but it seems to work for my applications at least. I have just modified the ParMesh to take a `create_gtopo` argument (defaults to `true`) and have the following at l218 of `pmesh.cpp`:

```
if (create_gtopo) { gtopo.Create(groups, 822); }
```

I've also made the `groups` ListOfIntegerSets variable a member of the class and added a (synchronizing) method `CreateGroupTopology` which can be called after the constructor is called asynchronously with `create_gtopo = false` on each process. Something like:

```
void ParMesh::CreateGroupTopology()
{
   // build the group communication topology
   MFEM_VERIFY(groups.Size() > 0, "Shared groups have not been correctly set up");
   gtopo.Create(groups, 822);
}
```

Let me know if you see any issues with this approach but otherwise I will just close the issue considering it resolved until a larger scale refactoring of the mesh partitioning process is available.
